src/arch/arm/64/machine/fpu.c

```diff
@@ -8,6 +8,9 @@
 #include <arch/machine/fpu.h>
 #include <mode/model/statedata.h>
 
+#define FPU_NOT_IMPLEMENTED (0xF)
+#define IS_FPU_ENABLED(id_aa64pfr0, shift) ((((id_aa64pfr0) >> (shift)) & MASK(4)) != FPU_NOT_IMPLEMENTED)
```

Comment:
It is a bit confusing if we use `MASK(4)` and `0xf` and `FPU_NOT_IMPLEMENTED`. Maybe simply say
```suggestion
#define IS_FPU_ENABLED(id_aa64pfr0, shift) ((((id_aa64pfr0) >> (shift)) & MASK(4)) != MASK(4))
```

Reply:
Or, looking at the ARM arch manual, make this really generic:

#define GET_ARM_CORE_FEATURE(id_aa64pfr, shift) ((id_aa64pfr) >> (shift)) & MASK(4))

#define GET_ARM_CORE_FEATURE_FPU(id_aa64pfr0) GET_ARM_CORE_FEATURE(id_aa64pfr0, ID_AA64PFR0_EL1_FP)
#define GET_ARM_CORE_FEATURE_ADVSIMD(id_aa64pfr0) GET_ARM_CORE_FEATURE(id_aa64pfr0, ID_AA64PFR0_EL1_ASIMD)

/* 0b1111 means "not implemented". All other values indicate the implementation type */
#define IS_FPU_ENABLED(id_aa64pfr0) (GET_ARM_CORE_FEATURE_FPU(id_aa64pfr0) != MASK(4))
#define IS_ADVSMDI_ENABLED(id_aa64pfr0) (GET_ARM_CORE_FEATURE_ADVSIMD(id_aa64pfr0) != MASK(4))
...
 if (IF_FPU_ENABLED(id_aa64pfr0) ||
        IS_ADVSMDI_ENABLED(id_aa64pfr0) {
        return true;
```